starter-code/routes/passportRouter.js

```diff
@@ -1,22 +1,69 @@
-const express        = require("express");
-const router         = express.Router();
+"use strict";
+
+const express = require("express");
+const router = express.Router();
 // User model
-const User           = require("../models/user");
+const User = require("../models/user");
 // Bcrypt to encrypt passwords
-const bcrypt         = require("bcrypt");
-const bcryptSalt     = 10;
+const bcrypt = require("bcrypt");
+const bcryptSalt = 10;
 const ensureLogin = require("connect-ensure-login");
-const passport      = require("passport");
+const passport = require("passport");
+const session    = require("express-session");
+
+
+
+
+router.get("/login", (req, res, next) => {
+  res.render("passport/login", {message: req.flash("error")});
+});
+
+router.post("/login", passport.authenticate("local", {
+  successRedirect: "/",
+  failureRedirect: "/login",
+  failureFlash: true,
+  passReqToCallback: true,
+  
+}));
 
 
 
 router.get("/private-page", ensureLogin.ensureLoggedIn(), (req, res) => {
-  res.render("passport/private", { user: req.user });
+  res.render("passport/private", {
+    user: req.user
+  });
 });
 
+router.get("/signup", (req, res) => {
+  res.render("passport/signup");
 
+});
+
+router.post("/signup", (req, res, next) => {
+  var username = req.body.username;
+  var password = req.body.password;
+  var salt = bcrypt.genSaltSync(bcryptSalt);
+  var hashPass = bcrypt.hashSync(password, salt);
+
+  var newUser = User({
+    username,
+    password: hashPass
+  });
+
+  if (username === "" || password === "") {
+    res.render("passport/signup", {
+      errorMessage: "Indicate a username and a password to sign up"
+    });
+    return;
+  }
 
```

Comment:
You are missing important validation here. Check that THE USER DOESN'T ALREADY EXISTS!!!:
```
User.findOne({ username }, "username", (err, user) => {
  if (user !== null) {
    res.render("passport/signup", { message: "The username already exists" });
    return;
  }

  var salt     = bcrypt.genSaltSync(bcryptSalt);
  var hashPass = bcrypt.hashSync(password, salt);

  var newUser = User({
    username,
    password: hashPass
  });

  newUser.save((err) => {
    if (err) {
      res.render("passport/signup", { message: "The username already exists" });
    } else {
      res.redirect("/login");
    }
  });
});
```